cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,755 @@
+    def from_adjacency_matrices(
+        adjacency_matrices: Dict[int, numpy.ndarray],
+        variable_names: Optional[List[Union[NodeLike, int]]] = None,
+    ) -> TimeSeriesCausalGraph:
+        """
+        Return a time series causal graph from a dictionary of adjacency matrices. Keys are the time delta.
+        This is useful for converting a list of adjacency matrices into a time series causal graph.
+
+        For example, the adjacency matrix with time delta -1 is stored in adjacency_matrices[-1] as would correspond to X-1 -> X,
+        where X is the set of nodes.
+
+        Example:
+        >>> adjacency_matrices = {
+        ...     -2: numpy.array([[0, 0, 0], [1, 0, 0], [0, 0, 1]]),
+        ...     -1: numpy.array([[0, 1, 0], [1, 0, 0], [0, 0, 0]]),
+        ...     0: numpy.array([[0, 1, 0], [0, 0, 1], [0, 1, 0]]),
+        ... }
+
+        Let the nodes be X,Y,Z. The corresponding edges are:
+        >>> edges = [
+        ...     (Z-2, Y),
+        ...     (X-1, X),
+        ...     (Y-1, X),
+        ...     (Y-1, Y),
+        ...     (Z-1, Z),
+        ...     (X, Z),
+        ...     (Z, Y),
+        ... ]
+
+        :param adjacency_matrices: A dictionary of adjacency matrices. Keys are the time delta.
+        :param variable_names: A list of variable names. If not provided, the variable names are integers starting from 0.
+            Node names must correspond to the variable names and must not contain the lag.
+        :return: A time series causal graph.
+        """
+        assert isinstance(adjacency_matrices, dict)
+        # keys must be integers or str that can be converted to integers
+        assert all(isinstance(key, (int, str)) and int(key) == key for key in adjacency_matrices)
+        if variable_names is not None:
+            variable_names_str: List[Union[str, int]] = []
+            assert len(variable_names) == adjacency_matrices[0].shape[0], (
+                'The number of variable names must be equal to the number of nodes in the adjacency matrix.'
+                f'Got {len(variable_names)} variable names and {adjacency_matrices[0].shape[0]} nodes.'
+            )
+            # convert the variable names to strings if they are not strings
+            for variable_name in variable_names:
+                if isinstance(variable_name, HasIdentifier):
+                    variable_names_str.append(variable_name.identifier)
+                else:
+                    variable_names_str.append(variable_name)
+
+        else:
+            variable_names_str = [f'node_{i}' for i in range(adjacency_matrices[0].shape[0])]
+
+        # we could create the full adjacency matrix from the adjacency matrices by stacking them according to the time delta
+        # but if we have many time deltas, this could be very memory intensive
+        # so we create the graph by adding the edges one by one for each time delta
+
+        # create the empty graph
+        tsgraph = TimeSeriesCausalGraph()
```

Comment:
I think it's better like this as now I changed the code to allow to possibility that time delta 0 is not present.

Reply:
I could do as you suggested for the first elements in the dict but I don't think it would change much. I will leave this comment open. If you want to change it, go for it :)

Reply:
Will leave as is as I agree that td 0 may not be present. Resolving this conversation.